core/src/main/java/io/grpc/internal/AbstractServerStream.java

```diff
@@ -243,7 +267,19 @@ public final void transportReportStatus(Status status) {
      * #transportReportStatus}.
      */
     public void complete() {
-      closeListener(Status.OK);
+      if (deframerClosed) {
+        deframerClosedTask = null;
+        closeListener(Status.OK);
+      } else {
+        deframerClosedTask =
+            new Runnable() {
+              @Override
+              public void run() {
+                closeListener(Status.OK);
+              }
+            };
+        closeDeframer(false);
```

Comment:
Done, and added test case to AbstractServerStreamTest. It seemed to fit better there than in the abstract transport test, as it's fairly low level.